ably/options_test.go

```diff
@@ -13,52 +13,65 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestDefaultFallbacks_RSC15h(t *testing.T) {
+func TestDefaultFallbacks_REC2c(t *testing.T) {
```

Comment:
I don't think we should override existing tests for deprecated clientOptions, instead they should be updated.
This is to make sure, deprecated features also work as expected.
 
Can you instead append new tests at the end of the test file and mark old tests as deprecated.
This is to make sure both old and new clientOptions work as expected. 

In the future, when we decide to remove deprecated clientOptions, we will remove both deprecated clientOptions and related tests.

Reply:
Unfortunately, the official [ably doc for clientOptions](https://ably.com/docs/api/realtime-sdk#client-options) and current README still recommends use of old clientOptions.

Reply:
Those docs won't be updated until we update all SDKs to use `endpoint` instead of `environment`

Reply:
> Can you instead append new tests at the end of the test file and mark old tests as deprecated.
This is to make sure both old and new clientOptions work as expected.

The changes here will change the behavour of the legacy clientOptions, so keeping the old tests as they were will mean they will fail. How should we handle this? 

> Unfortunately, the official [ably doc for clientOptions](https://ably.com/docs/api/realtime-sdk#client-options) and current README still recommends use of old clientOptions.

I can update the current README. For the SDK docs, this is the first client we are making this change on, and I am not an SDK developer, so will probably need some guidance and support for the process in updating docs, releasing new versions etc.

Reply:
Ideally, newly introduced `clientOptions` shouldn't break deprecated `clientOptions`.
Does new spec introduces breaking changes for deprecated clientOptions?
I would recommend to update the code in a way that, if `endpoint` is not specified, hosts will be calculated based on deprecated clientOptions

Reply:
Also, we have ongoing work for `chat`, `livesync`, updating all SDKs with `endpoint` behavior will take at least a year or so.. 
Also, not sure if we have internal priorities for updating clientOptions across SDKs.

Reply:
I am also adding @SimonWoolf to review the PR

Reply:
See the [discussion in the ADR](https://ably.atlassian.net/wiki/spaces/ENG/pages/3428810778/ADR-119+ClientOptions+for+new+DNS+structure#Issues-and-options) about updating the current behaviour of the clientOptions.

> Also, not sure if we have internal priorities for updating clientOptions across SDKs.

Yes, I guess that's why an SRE is doing this work rather than an SDK developer ;)

Reply:
Okay, ADR is still in progress 🤔, and seems decided to go with breaking option 2. 
Do we have the spec updated accordingly since ADR is still in progress?
If implemented with option2 ( also make sure we have spec updated accordingly ), we also need to update [UPDATING.md](https://github.com/ably/ably-go/blob/main/UPDATING.md ) and will mark a major release. i.e. 1.3.0

Reply:
The spec you linked to before does appear to be updated, yes. It describes the breaking change behaviour (that `environment` will become the routing policy name). 

> If implemented with option2 ( also make sure we have spec updated accordingly ), we also need to update [UPDATING.md](https://github.com/ably/ably-go/blob/main/UPDATING.md) and will mark a major release. i.e. 1.3.0

:+1: Is `1.3.0` a major release? I would have thought that would be `2.0.0`. Or is there some weird Ably context I'm not aware of :sweat_smile:

Reply:
Having discussed this with @paddybyers and @SimonWoolf, we don't necessarily think this is a breaking change. The tests _do_ need to change because the returned hostnames will change, but it's not actually causing any impact to the user.

The only thing that might cause a conflict is if both `WithEnvironment` and `WithRealtimeHost` or `WithRESTHost` are used together which previously was just ignored, but now will throw an error. I've updated the integration tests accordingly to use `WithEndpoint`. 

With this in mind, do we still mark this at 1.3.0? We don't actually specify clientOptions anywhere in docs as far as I can see, so not sure what I need to update, other than `CHANGELOG.md`.

Reply:
I agree, this is not a breaking change, at least not from an API perspective, no options have been removed, we're adding a new `endpoint` option and changing the hostnames that the `environment` option will now map to, which is not something we document nor guarantee to never change, so this can be a patch release.

> With this in mind, do we still mark this at 1.3.0? We don't actually specify clientOptions anywhere in docs as far as I can see, so not sure what I need to update, other than CHANGELOG.md.

I don't think we necessarily need to bump the minor version, I think the only thing we need to do is mark `WithEnvironment`, `WithRESTHost` and `WithRealtimeHost` as deprecated with a [deprecation comment](https://go.dev/wiki/Deprecated) saying "use WithEndpoint instead".

Reply:
Okay, [ADR](https://ably.atlassian.net/wiki/spaces/ENG/pages/3428810778/ADR-119+ClientOptions+for+new+DNS+structure#Issues-and-options) clearly states we planning to go with option 2 and it is a breaking change. 
Not sure, but there might be confusion here. If it's not a breaking change, can we update ADR + spec accordingly

Reply:
Also, if we planning to bump up the version to 1.3.*, we might as well add support for encryption as per latest issue reported by the customer -> https://ably-real-time.slack.com/archives/CURL4U2FP/p1736763296808579

Reply:
> If it's not a breaking change, can we update ADR + spec accordingly?

I'm going to follow up with @paddybyers to get the ADR updated to be clear that it is decided, and that this is not a breaking change. I will release this as the next minor version.